File: tags/meta/1438 longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.java
/*
https://leetcode.com/problems/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/

每一个时刻直到subarray中最大值和最小值才能知道当前的数组是否符合要求
需要两个单调栈，分别来追踪最大和最小值
每一时刻如果两个栈首差符合要求，那么右移窗口
如果不符合，需要右移窗口左边界
因为单调栈的元素都是顺序加入的，所以只有当栈首是当前左窗口的时候才需要移除，意为当前栈首已经不在考虑范围了

Time: O(n)
Space: O(n)
*/

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> minSt = new ArrayDeque<>(), maxSt = new ArrayDeque<>();
        int l = 0, r = 0, n = nums.length;
        while (r < n) {
            while (!minSt.isEmpty() && nums[minSt.peekLast()] >= nums[r]) minSt.pollLast();
            while (!maxSt.isEmpty() && nums[maxSt.peekLast()] <= nums[r]) maxSt.pollLast();
            minSt.addLast(r);
            maxSt.addLast(r);
            if (nums[maxSt.peekFirst()] - nums[minSt.peekFirst()] > limit) {
                if (maxSt.peekFirst() == l) maxSt.pollFirst();
                if (minSt.peekFirst() == l) minSt.pollFirst();
                l++;
            }
            r++;
        }
        return r - l;
    }
}

File: _archive/daily/1438 longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
'''
https://leetcode.com/problems/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/

sliding window变种
主要是要能快速知道区域内最大与最小值，且能在移动左边界时移除掉对应的值
所以维护两个单调栈，分别递增和递减
当入列右侧时，比它大/小的元素就不会对值有影响，可以pop出
移动左边界时，如果左边界恰好是栈首，那就移除掉即可

Time: O(n)
Space: O(n)
'''

class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        max_dq = deque()
        min_dq = deque()
        l = 0
        res = 0
        for r in range(len(nums)):
            while max_dq and max_dq[-1] < nums[r]:
                max_dq.pop()
            max_dq.append(nums[r])

            while min_dq and min_dq[-1] > nums[r]:
                min_dq.pop()
            min_dq.append(nums[r])

            while max_dq[0] - min_dq[0] > limit:
                if max_dq[0] == nums[l]:
                    max_dq.popleft()
                if min_dq[0] == nums[l]:
                    min_dq.popleft()
                l += 1
            res = max(res, r - l + 1)
        return res